syntax = "proto3";

package raft;

message AppendEntriesRequest {
    int32 requester_id = 1;
    int32 current_term = 2;
    int64 last_log_index = 4;
    int32 last_log_term = 5;
    repeated string op_entries = 6;
    int32 leader_commit_index = 7;

}

message AppendEntriesReply {
    int32 replier_id = 1;
    int32 current_term = 2;
    bool success = 3;
}

message VoteRequest{
    int32 requester_id = 1;
    int32 current_term = 2;
    int64 last_log_index = 3;
    int32 last_log_term = 4;
}

message VoteReply{
    int32 replier_id = 1;
    int32 current_term = 2;
    int64 last_log_index = 3;
    int32 last_log_term = 4;
    bool vote_granted = 5;
}

message SubmitRequest{
  int32 requester_id = 1;
}

message SubmitReply{
  int32 replier_id = 1;
  string message = 2;
}

service Raft{
    rpc AppendEntries(stream AppendEntriesRequest) returns (AppendEntriesReply);
    rpc RequestVote(VoteRequest) returns (VoteReply);
    rpc Submit(SubmitRequest) returns (SubmitReply);
}